datafusion/optimizer/src/reduce_cross_join.rs

```diff
@@ -0,0 +1,442 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Optimizer rule to reduce cross join to inner join if join predicates are available in filters.
+use crate::{OptimizerConfig, OptimizerRule};
+use datafusion_common::{Column, Result};
+use datafusion_expr::{
+    logical_plan::{Filter, Join, CrossJoin, JoinType, LogicalPlan},
+    utils::from_plan, and, or, utils::can_hash
+};
+use datafusion_expr::{Expr, Operator};
+
+use std::collections::HashSet;
+
+//use std::collections::HashMap;
+use std::sync::Arc;
+use datafusion_expr::logical_plan::JoinConstraint;
+
+#[derive(Default)]
+pub struct ReduceCrossJoin;
+
+impl ReduceCrossJoin {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for ReduceCrossJoin {
+    fn optimize(
+        &self,
+        plan: &LogicalPlan,
+        optimizer_config: &mut OptimizerConfig,
+    ) -> Result<LogicalPlan> {
+        let mut possible_join_keys : Vec<(Column, Column)> = vec![];
+        let mut all_join_keys = HashSet::new();
```

Comment:
It's a little odd that these are "global", not scoped with the recursion itself, but I think due to the way they are canonicalized, this is fine? I tried to break it in a few tests and couldn't. I think it would maybe break if there are multiple joins to the same tables aliased?

Reply:
I recommend adding some tests with three tables (not just 2)

Reply:
Yes, these variables need not be "Global", I am changing it to scope of the filter_expr.

Reply:
It seems like this particular comment is not addressed (they are still global). However, given the test coverage I think this PR can be merged as is

Reply:
@alamb This is addressed inside the reduce_cross_join(). New possible_join_keys are created for each Filter expression. This would be used only on it's child. 

```
fn reduce_cross_join(
    _optimizer: &ReduceCrossJoin,
    plan: &LogicalPlan,
    possible_join_keys: &mut Vec<(Column, Column)>,
    all_join_keys: &mut HashSet<(Column, Column)>,
) -> Result<LogicalPlan> {
    match plan {
        LogicalPlan::Filter(Filter { input, predicate }) => {
            // join keys are handled locally
            let mut new_possible_join_keys: Vec<(Column, Column)> = vec![];
            let mut new_all_join_keys = HashSet::new();
            extract_possible_join_keys(predicate, &mut new_possible_join_keys);
```